yt/config.py

```diff
@@ -66,68 +73,142 @@
+        node = None
+        use_fallback = "fallback" in kwargs
+        fallback = kwargs.pop("fallback", None)
+        while len(path) > 0:
+            try:
+                node = config
+                for k in path:
+                    node = node[k]
+                ok = True
+                break
+            except KeyError as e:
+                if strict and not use_fallback:
+                    raise e
+                else:
+                    path = path[:-1]
+
+        if not ok and use_fallback:
+            return fallback
+        elif not ok:
+            raise KeyError(f"Could not find {section}, {path} in configuration.")
+
+        return node
+
+    def update(self, new_values):
+        def copy_helper(dict_a, dict_b):
+            # Copies entries form dict_a in dict_b, inplace
+            for k in dict_a.keys():
+                if k not in dict_b:
+                    dict_b[k] = dict_a[k]
+                    continue
+                if isinstance(dict_a[k], dict):
+                    copy_helper(dict_a[k], dict_b[k])
+                else:
+                    dict_b[k] = dict_a[k]
+
+        copy_helper(new_values, self.values)
+
+    def has_section(self, section):
+        return section in self.values
+
+    def __setitem__(self, key, value):
+        section, *path, entry = key
+        if not self.has_section(section):
+            raise KeyError
+
+        node = self.values[section]
+        for p in path:
+            if p not in node:
+                node[p] = {}
+            node = node[p]
+
+        node[entry] = value
+
+    def read(self, file_names):
+        if not isinstance(file_names, (tuple, list)):
+            file_names = (file_names,)
+
+        file_names_read = []
+        for fname in file_names:
+            if not os.path.exists(fname):
+                continue
+            self.update(toml.load(fname))
+            file_names_read.append(fname)
 
-    def __getitem__(self, key):
-        self.get(key[0], key[1])
+        return file_names_read
 
-    def get(self, section, option, *args, **kwargs):
-        val = super(YTConfigParser, self).get(section, option, *args, **kwargs)
-        return os.path.expanduser(os.path.expandvars(val))
+    def write(self, fd):
+        config_as_str = toml.dumps(self.values)
+        try:
+            fd.write(config_as_str)
+        except AttributeError:
+            with open(fd, mode="w") as fdd:
+                fdd.write(config_as_str)
 
 
-ytcfg = YTConfigParser(ytcfg_defaults)
-ytcfg.read([_OLD_CONFIG_FILE, CURRENT_CONFIG_FILE, "yt.cfg"])
-if not ytcfg.has_section("yt"):
-    ytcfg.add_section("yt")
+# Walk the tree up until we find a config file
+ytcfg = YTConfig(ytcfg_defaults)
+if os.path.exists(CURRENT_CONFIG_FILE):
+    ytcfg.read(CURRENT_CONFIG_FILE)
+cwd = Path(".").absolute()
+while cwd.parent != cwd:
+    cfg_file = cwd / "yt.toml"
+    if cfg_file.exists():
+        ytcfg.read(cfg_file)
+        break
+    cwd = cwd.parent
```

Comment:
I'm looking at this and trying to figure out if it's possible to get to the root, or to get into an infinite loop, and I can't quite tell.  Is it ever possible we'll see an exception for trying to get to the parent of a non-parented path?

Reply:
I _think_ the code is correct, but I'll let others weigh in.

Reply:
I updated this bit as it was causing me headache to make sure it will work in *any* case.